OSSM-8190: Add product documentation for Multi-Cluster Topology (#84640)
openshift/openshift-docs: modules/ossm-applying-certificates-to-multi-cluster-topology.adoc

:_mod-docs-content-type: Procedure
[id="ossm-applying-certificates-to-multi-cluster-topology_{context}"]
= Applying certificates to a multi-cluster topology 

> This procedure explains how to apply security certificates to a multi-cluster topology.

Author: done

.Prerequisites

* You have access to two {ocp-product-title} clusters with external load balancer support. 

> * You have installed the {SMProduct} 3.0 Operator to two clusters.

Reviewer: Instead of "two", should this refer to the two clusters mentioned in the previous bullet? Something like this:
```suggestion
* You have installed the {SMProduct} 3.0 Operator to these clusters.
```

Author: done

* You have created the root CA certificate and intermediate CA certificates for each cluster or someone has made them available for you. 

[NOTE]
====

> In this procedure, `cluster1` is the East cluster and `cluster2` is the West cluster. 

Reviewer: Should this be in  uppercase (`CLUSTER1` and `CLUSTER2`), since the only time we see these terms are in the environment variable names (`CTX_CLUSTER1`, `CTX_CLUSTER2`)?

Author: done

====

.Procedure

> . Log in to the first cluster by running the following command:

Reviewer: Maybe use "east cluster" instead of "first cluster" throughout the entire doc, since the doc starts off with a note that mentiones east and west instead of first and second cluster.

Author: done

+
[source,terminal]
----
$ $ oc login -u https://<first_cluster_api_server_url>
----

> . Set up the environment variable that contains the context for the first cluster by running the following command:

Reviewer: ```suggestion
. Set up the environment variable that contains the `oc` command context for the first cluster by running the following command:
```

Author: done

+
[source,terminal]
----
$ export CTX_CLUSTER1=$(oc config current-context)
----

. Create a namespace called `istio-system` by running the following command:
+
[source,terminal]
----

> $ oc get ns istio-system --context "${CTX_CLUSTER1}" || oc create namespace istio-system --context "${CTX_CLUSTER1}"

Reviewer: We may need to replace `namespace` with `project` here, since we're talking about OpenShift, not vanilla Kubernetes.
```suggestion
$ oc get project istio-system --context "${CTX_CLUSTER1}" || oc new-project istio-system --context "${CTX_CLUSTER1}"
```

Author: done

----

. Configure {istio} to use `network1` as the default network for the pods on the East cluster by running the following command:
+
[source,terminal]
----
oc --context "${CTX_CLUSTER1}" label namespace istio-system topology.istio.io/network=network1
----

> . Create the CA certificate secret for {istio} on the East cluster by running the following command:

Reviewer: To be precise, the secret contains two CA certificates, a certificate chain, and a private key, so "CA certificate secret" may not be completely correct.

Author: done

+
[source,terminal]
----
oc get secret -n istio-system --context "${CTX_CLUSTER1}" cacerts || oc create secret generic cacerts -n istio-system --context "${CTX_CLUSTER1}" \
  --from-file=east/ca-cert.pem \
  --from-file=east/ca-key.pem \
  --from-file=east/root-cert.pem \
  --from-file=east/cert-chain.pem
----
+
[NOTE]
====

> If you followed the instructions in _Creating certificates for a multi-cluster mesh_ then your certificates will reside in the directory specified in the previous example. If your certificates reside in a different directory, modify the syntax accordingly. 

Reviewer: Not sure what "previous example" refers to here. Maybe call out the `east/` directory specifically?
```suggestion
If you followed the instructions in _Creating certificates for a multi-cluster mesh_ then your certificates will reside in the `east/` directory. If your certificates reside in a different directory, modify the syntax accordingly. 
```

Author: done

====

> 

Reviewer: This is a good place to split the procedure into two sections, so you'll have two sections of 5 steps each instead of one section with 10 steps. The steps for cluster1 end here; steps for cluster2 follow from here till the end of the document.

Author: done, but let's look at this and discus it.

Author: done

. Log in to the second cluster by running the following command:
+
[source,terminal]
----

> $ $ oc login -u https://<second_cluster_api_server_url>

Reviewer: ```suggestion
oc login -u https://<second_cluster_api_server_url>
```

Author: done

----

> . Set up the environment variable that contains the context for the second cluster by running the following command:

Reviewer: ```suggestion
. Set up the environment variable that contains the `oc` command context for the second cluster by running the following command:
```

Author: done

+
[source,terminal]
----
$ export CTX_CLUSTER1=$(oc config current-context)
----

. Create a namespace called `istio-system` by running the following command:
+
[source,terminal]
----

> $ oc get ns istio-system --context "${CTX_CLUSTER2}" || oc create namespace istio-system --context "${CTX_CLUSTER2}"

Reviewer: ```suggestion
$ oc get project istio-system --context "${CTX_CLUSTER2}" || oc new-project istio-system --context "${CTX_CLUSTER2}"
```

Author: done

----

. Configure {istio} to use `network2` as the default network for the pods on the West cluster by running the following command:
+
[source,terminal]
----
oc --context "${CTX_CLUSTER2}" label namespace istio-system topology.istio.io/network=network2
----

. Create the CA certificate secret for {istio} on the West cluster by running the following command:
+
[source,terminal]
----
oc get secret -n istio-system --context "${CTX_CLUSTER2}" cacerts || oc create secret generic cacerts -n istio-system --context "${CTX_CLUSTER2}" \
  --from-file=west/ca-cert.pem \
  --from-file=west/ca-key.pem \
  --from-file=west/root-cert.pem \
  --from-file=west/cert-chain.pem
----
+
[NOTE]
====
If you followed the instructions in _Creating certificates for a multi-cluster mesh_ then your certificates will reside in the directory specified in the previous example. If the certificates reside in a different directory, modify the syntax accordingly. 
====

.Next steps

> Install the Operator. 

Reviewer: Add name of the operator and that it has to be installed on both clusters.
```suggestion
Install the Service Mesh Operator on both clusters. 
```

Reviewer: Also, maybe installing the operator could be the first step, so you get this sequence:
1. install operator on both clusters
2. create the certificates for both clusters
3. apply the certificates to both clusters
4. install istio on both clusters

This way, the "Next step" in this document would be more natural, since you need the certificates for Istio, not for the operator.

// This module is used in the following assemblies:

// * install/ossm-multi-cluster-topology.adoc

> :_mod-docs-content-type: Procedure

Reviewer 2: ```suggestion
:_mod-docs-content-type: PROCEDURE
```

Author: done

[id="ossm-applying-certificates-to-multi-cluster-topology_{context}"]
= Applying certificates to a multi-cluster topology 

Apply root and intermediate certificate authority (CA) certificates to the clusters in a multi-cluster topology.

[NOTE]
====
In this procedure, `CLUSTER1` is the East cluster and `CLUSTER2` is the West cluster. 
====

.Prerequisites

* You have access to two {ocp-product-title} clusters with external load balancer support. 

* You have created the root CA certificate and intermediate CA certificates for each cluster or someone has made them available for you. 

.Procedure

> . Apply the certificates to the East cluster of the multi-cluster topology.

Reviewer 2: ISG: Colons leading into substeps:
```suggestion
. Apply the certificates to the East cluster of the multi-cluster topology:
```

Author: done

.. Log in to East cluster by running the following command:
+
[source,terminal]
----
$ oc login -u https://<east_cluster_api_server_url>
----

.. Set up the environment variable that contains the `oc` command context for the East cluster by running the following command:
+
[source,terminal]
----
$ export CTX_CLUSTER1=$(oc config current-context)
----

> .. Create a namespace called `istio-system` by running the following command:

Author: done

+
[source,terminal]
----

> $ oc get project istio-system --context "${CTX_CLUSTER1}" || oc  new-project istio-system --context "${CTX_CLUSTER1}"

Reviewer 2: Extra space:
```suggestion
$ oc get project istio-system --context "${CTX_CLUSTER1}" || oc new-project istio-system --context "${CTX_CLUSTER1}"
```

Author: done

> $ oc get project istio-system --context "${CTX_CLUSTER1}" || oc create new-project istio-system --context "${CTX_CLUSTER1}"

Reviewer 3: `$ oc get project istio-system --context "${CTX_CLUSTER1}" || oc  new-project istio-system --context "${CTX_CLUSTER1}"`

update the command

Author: done

----

.. Configure {istio} to use `network1` as the default network for the pods on the East cluster by running the following command:
+
[source,terminal]
----
$ oc --context "${CTX_CLUSTER1}" label namespace istio-system topology.istio.io/network=network1
----

.. Create the CA certificates, certificate chain, and the private key for {istio} on the East cluster by running the following command:
+
[source,terminal]
----
$ oc get secret -n istio-system --context "${CTX_CLUSTER1}" cacerts || oc create secret generic cacerts -n istio-system --context "${CTX_CLUSTER1}" \
  --from-file=east/ca-cert.pem \
  --from-file=east/ca-key.pem \
  --from-file=east/root-cert.pem \
  --from-file=east/cert-chain.pem
----
+
[NOTE]
====

> If you followed the instructions in _Creating certificates for a multi-cluster mesh_ your certificates will reside in the `east/` directory. If your certificates reside in a different directory, modify the syntax accordingly. 

Reviewer 2: Quotations for section references ([repo guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#lead-in-link-sentence-style-for-non-xref-situations)), and a comma:
```suggestion
If you followed the instructions in "Creating certificates for a multi-cluster mesh", your certificates will reside in the `east/` directory. If your certificates reside in a different directory, modify the syntax accordingly. 
```

Author: done

====

> . Apply the certificates to the West cluster of the multi-cluster topology.

Reviewer 2: ```suggestion
. Apply the certificates to the West cluster of the multi-cluster topology:
```

Author: done

.. Log in to the West cluster by running the following command:
+
[source,terminal]
----
$ oc login -u https://<west_cluster_api_server_url>
----

.. Set up the environment variable that contains the `oc` command context for the West cluster by running the following command:
+
[source,terminal]
----

> $ export CTX_CLUSTER1=$(oc config current-context)

Reviewer 3: `export CTX_CLUSTER2=$(oc config current-context)`

update this command from cluster1 to cluster2

Author: Nice catch! Thanks, made the change.

----

> .. Create a namespace called `istio-system` by running the following command:

Author: done

+
[source,terminal]
----
$ oc get project istio-system --context "${CTX_CLUSTER2}" || oc new-project istio-system --context "${CTX_CLUSTER2}"
----

.. Configure {istio} to use `network2` as the default network for the pods on the West cluster by running the following command:
+
[source,terminal]
----

> oc --context "${CTX_CLUSTER2}" label namespace istio-system topology.istio.io/network=network2

Reviewer 2: Command prompt:
```suggestion
$ oc --context "${CTX_CLUSTER2}" label namespace istio-system topology.istio.io/network=network2
```

Author: done

----

.. Create the CA certificate secret for {istio} on the West cluster by running the following command:
+
[source,terminal]
----

> oc get secret -n istio-system --context "${CTX_CLUSTER2}" cacerts || oc create secret generic cacerts -n istio-system --context "${CTX_CLUSTER2}" \

Reviewer 2: ```suggestion
$ oc get secret -n istio-system --context "${CTX_CLUSTER2}" cacerts || oc create secret generic cacerts -n istio-system --context "${CTX_CLUSTER2}" \
```

Author: done

  --from-file=west/ca-cert.pem \
  --from-file=west/ca-key.pem \
  --from-file=west/root-cert.pem \
  --from-file=west/cert-chain.pem
----
+
[NOTE]
====

> If you followed the instructions in _Creating certificates for a multi-cluster mesh_ your certificates will reside in the `west/` directory. If the certificates reside in a different directory, modify the syntax accordingly. 

Reviewer 2: ```suggestion
If you followed the instructions in "Creating certificates for a multi-cluster mesh", your certificates will reside in the `west/` directory. If the certificates reside in a different directory, modify the syntax accordingly. 
```

Author: done

====

.Next steps

> Install {istio} on both clusters. See _Installing a multi-primary multi-network mesh_ for more information.

Author: Given that we only have the one procedure at this time, but we will need 4 for GA, I'm thinking I should either just keep this generic and say "Install {istio} on all the clusters comprising the mesh".

What do you think @luksa? Could that work for now?

> Install {istio} on both clusters. See _Installing a multi-primary multi-network mesh_ for more information.

Author: done
